core/rawdb/accessors_chain.go

```diff
@@ -173,18 +173,33 @@ func WriteFastTrieProgress(db ethdb.KeyValueWriter, count uint64) {
 
 // ReadHeaderRLP retrieves a block header in its raw RLP database encoding.
 func ReadHeaderRLP(db ethdb.Reader, hash common.Hash, number uint64) rlp.RawValue {
+	// First try to look up the data in ancient database. Extra hash
+	// comparison is necessary since ancient database only maintains
+	// the canonical data.
 	data, _ := db.Ancient(freezerHeaderTable, number)
-	if len(data) == 0 {
-		data, _ = db.Get(headerKey(number, hash))
-		// In the background freezer is moving data from leveldb to flatten files.
-		// So during the first check for ancient db, the data is not yet in there,
-		// but when we reach into leveldb, the data was already moved. That would
-		// result in a not found error.
-		if len(data) == 0 {
-			data, _ = db.Ancient(freezerHeaderTable, number)
+	if len(data) > 0 {
+		h, _ := db.Ancient(freezerHashTable, number)
```

Comment:
the data we have in `h` right now, I guess we could also just `hash` that and not lookup from disk? Not sure what's best, just mentioning the alternative...

Reply:
Sorry, I don't understand. What do you mean we don't need to lookup from disk?

Reply:
Like this: 
```diff
diff --git a/core/rawdb/accessors_chain.go b/core/rawdb/accessors_chain.go
index 2a5c72dedc..425941c358 100644
--- a/core/rawdb/accessors_chain.go
+++ b/core/rawdb/accessors_chain.go
@@ -23,6 +23,7 @@ import (
 
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
 	"github.com/ethereum/go-ethereum/ethdb"
 	"github.com/ethereum/go-ethereum/log"
 	"github.com/ethereum/go-ethereum/params"
@@ -178,7 +179,7 @@ func ReadHeaderRLP(db ethdb.Reader, hash common.Hash, number uint64) rlp.RawValu
 	// the canonical data.
 	data, _ := db.Ancient(freezerHeaderTable, number)
 	if len(data) > 0 {
-		h, _ := db.Ancient(freezerHashTable, number)
+		h := crypto.Keccak256(data)
 		if common.BytesToHash(h) == hash {
 			return data
 		}
@@ -194,7 +195,7 @@ func ReadHeaderRLP(db ethdb.Reader, hash common.Hash, number uint64) rlp.RawValu
 	// result in a not found error.
 	data, _ = db.Ancient(freezerHeaderTable, number)
 	if len(data) > 0 {
-		h, _ := db.Ancient(freezerHashTable, number)
+		h := crypto.Keccak256(data)
 		if common.BytesToHash(h) == hash {
 			return data
 		}

```

Reply:
I would think hashing this on the fly would be "cheaper" than hitting disk, in most cases. Headers aren't that big..